pkg/keylock/keylock.go

```diff
@@ -83,6 +92,31 @@ func (kl *keyLock) Acquire(id string) LockHandle {
 	return *h
 }
 
+func (kl *keyLock) AcquireWithTimeout(id string, duration time.Duration) (LockHandle, error) {
+	h := kl.getOrCreateLock(id)
+	timeout := time.After(duration)
+	for {
+		select {
+		case <-timeout:
+			h.refcnt--
+			return *h, fmt.Errorf("error waiting for lock")
+		default:
+			if kl.tryLock(id) {
+				h.genNum++
+				return *h, nil
+			}
+			time.Sleep(1 * time.Millisecond)
+		}
+	}
+
+}
+
+func (kl *keyLock) tryLock(id string) bool {
+	h := kl.getOrCreateLock(id)
```

Comment:
This will increase refcnt every time trylock is called? We should probably use the same h that AcquireWithTimeout has.

Reply:
fixed.